internal/charmstore/migrations_dump_test.go

```diff
@@ -0,0 +1,715 @@
+// can be used to backtrack the database to any of the dumped versions.
+func dumpMigrationHistory(session *mgo.Session, earliestDeployedVersion string, history []versionSpec) error {
+	db := session.DB(historicalDBName)
+	vcsStatus, err := currentVCSStatus()
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	dumping := false
+	for _, vc := range history {
+		logger.Infof("----------------- running version %v", vc.version)
+		if vc.version == earliestDeployedVersion {
+			dumping = true
+		}
+		if err := runMigrationVersion(db, vc); err != nil {
+			return errgo.Notef(err, "cannot run at version %s", vc.version)
+		}
+		if dumping {
+			filename := migrationDumpFileName(vc.version)
+			logger.Infof("dumping database to %s", filename)
+			if err := saveDBToFile(db, vcsStatus, filename); err != nil {
+				return errgo.Notef(err, "cannot save DB at version %v", vc.version)
+			}
+		}
+	}
+	if !dumping {
+		return errgo.Newf("no versions matched earliest deployed version %q; nothing dumped", earliestDeployedVersion)
+	}
+	return nil
+}
+
+// createDatabaseAtVersion changes the state of the
+// database to be as it is after running the database
+// migrations from all the versions up to and including
+// the given version, which must be a version in the
+// dumped history not before the earliest deployed version.
+//
+// It finds the database snapshot from the appropriate
+// dump file (see dumpMigrationHistory).
+func createDatabaseAtVersion(db *mgo.Database, version string) error {
+	vcsStatus, err := restoreDBFromFile(db, migrationDumpFileName(version))
+	if err != nil {
+		return errgo.Notef(err, "cannot restore version %q", version)
+	}
+	logger.Infof("restored migration from version %s; dumped at %s", version, vcsStatus)
+	return nil
+}
+
+// migrationDumpFileName returns the name of the file that
+// the migration database snapshot will be saved to.
+func migrationDumpFileName(version string) string {
+	return "migrationdump." + version + ".zip"
+}
+
+// currentVCSStatus returns the git status of the current
+// charmstore source code. This will be saved into the
+// migration dump file so that there is some indication
+// as to when that was created.
+func currentVCSStatus() (string, error) {
+	cmd := exec.Command("git", "describe")
+	cmd.Stderr = os.Stderr
+	data, err := cmd.Output()
+	if err != nil {
+		return "", errgo.Mask(err)
+	}
+	// With the --porcelain flag, git status prints a simple
+	// line-per-locally-modified-file, or nothing at all if there
+	// are no locally modified files.
+	cmd = exec.Command("git", "status", "--porcelain")
+	cmd.Stderr = os.Stderr
+	data1, err := cmd.Output()
+	if err != nil {
+		return "", errgo.Mask(err)
+	}
+	return string(append(data, data1...)), nil
+}
+
+// saveDBToFile dumps the entire state of the database to the given
+// file name, also saving the given VCS status.
+func saveDBToFile(db *mgo.Database, vcsStatus string, filename string) (err error) {
```

Comment:
Using mongodump would create a directory that we'd have to zip if we wanted the nice neat dump file. We'd have less control over what we dumped (for example, we deliberately don't dump the indexes because they're  created anew each time the server runs). Also we'd still have to use custom restore code because I don't want the usual test code to rely on extra tools that might not be around. If we're doing the restore ourselves, it seems reasonable to do the dump so that we're sure about what we're doing.